x/pubkey/keeper/msg_server.go

```diff
@@ -39,22 +39,24 @@ func (m msgServer) AddKey(goCtx context.Context, msg *types.MsgAddKey) (*types.M
 		return nil, types.ErrValidatorNotFound
 	}
 
-	pubKey, ok := msg.PubKey.GetCachedValue().(cryptotypes.PubKey)
-	if !ok {
-		return nil, sdkerrors.ErrInvalidType.Wrapf("%T is not a cryptotypes.PubKey", pubKey)
+	for _, indPubKey := range msg.IndexedPubKeys {
+		pubKey, ok := indPubKey.PubKey.GetCachedValue().(cryptotypes.PubKey)
+		if !ok {
+			return nil, sdkerrors.ErrInvalidType.Wrapf("%T is not a cryptotypes.PubKey", pubKey)
+		}
+
+		if err := m.Keeper.PubKeys.Set(ctx, collections.Join(valAddr, indPubKey.Index), pubKey); err != nil {
+			return nil, err
+		}
```

Comment:
For a different PR: move the knowledge of the store/collections out of the message server?